common/kafka/kafka_consumer.cpp

```diff
@@ -20,293 +20,248 @@
-      topic_names_string_.append(", ");
+        tmp_topic_partitions.push_back(topic_partitions.back().get());
+        topic_names_string_.append(topic_name);
+        topic_names_string_.append(", ");
+      }
     }
-  }
-  const auto error_code = ExecuteKafkaOperationWithRetry(
+    const auto error_code = ExecuteKafkaOperationWithRetry(
       [this, &tmp_topic_partitions]() {
-        return consumer_->assign(tmp_topic_partitions); });
-  if (error_code != RdKafka::ERR_NO_ERROR) {
-    LOG(ERROR) << "Failed to assign partitions: " << partition_ids_str_
-               << ", error_code: " << RdKafka::err2str(error_code)
-               << ", num retries: " << FLAGS_kafka_consumer_num_retries;
-    common::Stats::get()->Incr(getFullStatsName(
+        return consumer_->getInstance()->assign(tmp_topic_partitions);
+      });
+
+    if (error_code != RdKafka::ERR_NO_ERROR) {
+      LOG(ERROR) << "Failed to assign partitions: " << partition_ids_str_
+                 << ", error_code: " << RdKafka::err2str(error_code)
+                 << ", num retries: " << FLAGS_kafka_consumer_num_retries;
+      common::Stats::get()->Incr(getFullStatsName(
         kKafkaConsumerErrorAssign,
         {kafka_consumer_type_metric_tag_,
          "error_code=" + std::to_string(error_code)}));
-    return;
+
+      if (std::atomic_load(&reset_)) {
+        continue;
+      }
+
+      return;
+    }
+    LOG(INFO) << "Successfully created kafka consumer for partitions: "
+              << partition_ids_str_ << ", topics: "
+              << folly::join(",", topic_names_);
+    is_healthy_.store(true);
+  } while (std::atomic_load(&reset_));
+}
+
+KafkaConsumer::~KafkaConsumer() {
+  if (!FLAGS_kafka_consumer_reset_on_file_change.empty()) {
+    if (cbIdPtr_ != nullptr) {
+      const auto cbId = *cbIdPtr_.get();
+      SSLFilePollerHolder::getFilePollerInstance()->cancelCallback(cbId);
+    }
+  }
+
+  if (consumer_) {
+    consumer_->close();
   }
-  LOG(INFO) << "Successfully created kafka consumer for partitions: "
-            << partition_ids_str_ << ", topics: "
-            << folly::join(",", topic_names_);
-  is_healthy_.store(true);
 }
 
 bool KafkaConsumer::IsHealthy() const { return is_healthy_.load(); }
 
 bool KafkaConsumer::Seek(const int64_t timestamp_ms) {
-  if (SeekInternal(topic_names_, timestamp_ms)) {
-    return true;
-  } else {
-    common::Stats::get()->Incr(
+  do {
+    while (std::atomic_exchange(&reset_, false)) {
+      consumer_->resetInstance(nullptr);
+    }
+
+    if (SeekInternal(topic_names_, timestamp_ms)) {
+      return true;
+    } else {
+      common::Stats::get()->Incr(
         getFullStatsName(kKafkaConsumerErrorSeek,
-            {kafka_consumer_type_metric_tag_}));
-    return false;
-  }
+                         {kafka_consumer_type_metric_tag_}));
+
+      if (std::atomic_load(&reset_)) {
+        continue;
+      }
+      return false;
+    }
+  } while (std::atomic_load(&reset_));
```

Comment:
Do we need a while loop here? If the flag gets reset while seeking (line 191), isn't it okay to return false? The caller can retry.

Reply:
call sites can be at different places... Isn't it better to retry here to ensure all callsites will get affected and it is transparent to call sites?